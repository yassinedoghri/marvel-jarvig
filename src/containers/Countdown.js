import PropTypes from "prop-types";
import React, { Component } from "react";

class Countdown extends Component {
  static secondsToTime(secs) {
    const hours = Math.floor(secs / (60 * 60));

    const divisorForMinutes = secs % (60 * 60);
    const minutes = Math.floor(divisorForMinutes / 60);

    const divisorForSeconds = divisorForMinutes % 60;
    const seconds = Math.ceil(divisorForSeconds);

    return {
      h: hours,
      m: minutes,
      s: seconds
    };
  }

  constructor(props) {
    super(props);

    this.state = {
      time: Countdown.secondsToTime(props.from),
      seconds: props.from
    };
    this.timer = 0;
    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
    this.countDown = this.countDown.bind(this);
  }

  componentDidMount() {
    this.startTimer();
  }

  componentDidUpdate() {
    const { isPaused, onCountdownPaused } = this.props;
    if (isPaused) {
      onCountdownPaused(this.state.time);
    }
  }

  componentWillUnmount() {
    this.stopTimer();
  }

  startTimer() {
    if (this.timer === 0) {
      this.timer = setInterval(this.countDown, 1000);
    }
  }

  stopTimer() {
    if (this.timer) {
      clearInterval(this.timer);
    }
    this.timer = 0;
  }

  countDown() {
    // Remove one second, set state so a re-render happens.
    const { isPaused, onCountdownEnd } = this.props;

    if (!isPaused) {
      const seconds = this.state.seconds - 1;
      this.setState({
        time: Countdown.secondsToTime(seconds),
        seconds
      });

      // Check if we're at zero.
      if (seconds === 0) {
        this.stopTimer();
        onCountdownEnd(this.state.time);
      }
    }
  }

  render() {
    const { time } = this.state;
    const timer =
      time.m === 0
        ? time.s
        : `${time.m}:${time.s < 10 ? `0${time.s}` : time.s}`;

    return <React.Fragment>{timer}</React.Fragment>;
  }
}

Countdown.propTypes = {
  from: PropTypes.number.isRequired,
  isPaused: PropTypes.bool,
  onCountdownPaused: PropTypes.func,
  onCountdownEnd: PropTypes.func
};

Countdown.defaultProps = {
  isPaused: false,
  onCountdownPaused: () => {},
  onCountdownEnd: () => {}
};

export default Countdown;
