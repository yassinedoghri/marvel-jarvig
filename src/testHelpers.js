import React from "react";
import "jest-styled-components";
import renderer from "react-test-renderer";
import theme from "theme";
import { ThemeProvider } from "styled-components";

export default function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}
