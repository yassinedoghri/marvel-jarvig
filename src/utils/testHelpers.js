import { createMemoryHistory } from "history";
import React from "react";
import { Provider } from "react-redux";
import {
  ConnectedRouter,
  routerMiddleware,
  routerReducer
} from "react-router-redux";
import renderer from "react-test-renderer";
import ShallowRenderer from "react-test-renderer/shallow";
import rootReducer from "reducers";
import { applyMiddleware, combineReducers, createStore } from "redux";
import { ThemeProvider } from "styled-components";
import theme from "theme";

const history = createMemoryHistory("/");
const routeMiddleware = routerMiddleware(history);

const store = createStore(
  combineReducers({
    ...rootReducer,
    router: routerReducer
  }),
  applyMiddleware(routeMiddleware)
);

export function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}

export function renderInContext(component) {
  return renderer.create(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={theme}>{component}</ThemeProvider>
      </ConnectedRouter>
    </Provider>
  );
}

export function ShallowRenderInContext(component) {
  const renderer = new ShallowRenderer();
  return renderer.render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <ThemeProvider theme={theme}>{component}</ThemeProvider>
      </ConnectedRouter>
    </Provider>
  );
}
