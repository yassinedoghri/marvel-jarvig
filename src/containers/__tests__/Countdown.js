import React from "react";
import { renderInContext } from "testHelpers";
import Countdown from "containers/Countdown";

describe("Countdown container", () => {
  it("renders without crashing", () => {
    renderInContext(
      <Countdown
        result={[]}
        from={2}
        onCountdownEnd={() => console.info("test")}
        isGamePaused={false}
      />
    );
  });
});
