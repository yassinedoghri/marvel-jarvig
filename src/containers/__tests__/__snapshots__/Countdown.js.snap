// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown container renders without crashing 1`] = `"2:00"`;
